ui/org.eclipse.pde.launching/src/org/eclipse/pde/launching/JUnitEclipsePluginValidationOperation.java

```diff
@@ -0,0 +1,93 @@
+package org.eclipse.pde.launching;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.debug.core.ILaunchConfiguration;
+import org.eclipse.pde.core.plugin.IPluginModelBase;
+import org.eclipse.pde.internal.launching.launcher.EclipsePluginValidationOperation;
+import org.osgi.framework.Version;
+import org.osgi.framework.hooks.resolver.ResolverHook;
+import org.osgi.framework.wiring.BundleCapability;
+import org.osgi.framework.wiring.BundleRequirement;
+import org.osgi.framework.wiring.BundleRevision;
+
+final class JUnitEclipsePluginValidationOperation extends EclipsePluginValidationOperation {
+
+	private final JunitValidationHook junitValidationHook;
+
+	public JUnitEclipsePluginValidationOperation(ILaunchConfiguration configuration, Set<IPluginModelBase> models) {
+		super(configuration, models);
+		junitValidationHook = new JunitValidationHook(configuration);
+		if (junitValidationHook.junitVersion == 5) {
+			setResolverHook(junitValidationHook);
+		}
+	}
+
+	@Override
+	public boolean hasErrors() {
+		return super.hasErrors() || !junitValidationHook.errors.isEmpty();
+	}
+
+	@Override
+	public Map<Object, Object[]> getInput() {
+		Map<Object, Object[]> map = super.getInput();
+		map.putAll(junitValidationHook.errors);
+		return map;
+	}
+
+	private static class JunitValidationHook implements ResolverHook {
+
+		private final Map<Object, Object[]> errors;
+		private final int junitVersion;
+
+		@SuppressWarnings("restriction")
+		private JunitValidationHook(ILaunchConfiguration configuration) {
+			errors = new HashMap<>(2);
+			org.eclipse.jdt.internal.junit.launcher.ITestKind testKind = org.eclipse.jdt.internal.junit.launcher.JUnitLaunchConfigurationConstants.getTestRunnerKind(configuration);
+			if (org.eclipse.jdt.internal.junit.launcher.TestKindRegistry.JUNIT5_TEST_KIND_ID.equals(testKind.getId())) {
+				junitVersion = 5;
+			} else {
+				junitVersion = -1;
+			}
+		}
+
+		@Override
+		public void filterMatches(BundleRequirement requirement, Collection<BundleCapability> candidates) {
+			Map<String, String> directives = requirement.getDirectives();
+			if (!"optional".equals(directives.get("resolution"))) { //$NON-NLS-1$ //$NON-NLS-2$
+				String requirementName = requirement.getRevision().getSymbolicName();
+				if (!requirementName.startsWith("junit")) { //$NON-NLS-1$
+					Iterator<BundleCapability> iterator = candidates.iterator();
+					while (iterator.hasNext()) {
+						BundleCapability candidate = iterator.next();
+						BundleRevision candidateRevision = candidate.getRevision();
+						String name = candidateRevision.getSymbolicName();
+						Version version = candidateRevision.getVersion();
+						if (version.getMajor() > junitVersion && name.startsWith("junit")) { //$NON-NLS-1$
```

Comment:
Technically combinations of 5 and 6 are a problem, which is not covered well by `version.getMajor() > junitVersion`. But we can change this once JUnit 6 support is there.

Reply:
> Technically combinations of 5 and 6 are a problem, which is not covered well by version.getMajor() > junitVersion. But we can change this once JUnit 6 support is there.

Let us first cover one thing at a time, the actual technical background does not matter so mutch but we should let the user knwo it uses a different version than the one is selected to launch with. SO at best we would use version ranges for Junit 5 + 6 and ... while 5 is a bit messy because it has different ones with 1.x and 5.x